# Automagic Phylogenies

## Description 

This project consists in a Nextflow pipeline that executes the entire process of a phylogenetic analysis, from the biological data extraction to the creation of the phylogenetic tree.

In order to proceed with the pipeline, the following tools were used:

- EntrezAPI Biopython package: Sequence extraction;
- KAlign2: Alignment of the biological sequences;
- AMAS: Sequence concatenation;
- Modeltest-NG: Best evolutionary model determination;
- RAxML-NG: Maximum-likelihood phylogenetic tree construction;
- Toytree & Toyplot: Manipulation and format convertion of the phylogenetic tree.

## Requirements
- Docker v28.0.1
- Nextflow v24.10.4

## Usage
Before running the pipeline, don't forgot to give give yourself permission to use Docker.

Try this:
```
sudo usermod -aG docker $USER
newgrp docker
```

The pipeline must be runned with a Docker image.

In order to create one try this:
```
docker build -t {image name} .
```
### Example
``` 
docker build -t automagic .
```

Now that you have a Docker image, you can run the pipeline with no issues:
```
nextflow pipeline.nf --o "{organism}" --t "{taxonomic level}" --docker "{Docker image}" --outgroup "{outgroup organism}"(optional)
```

### Example 
```
nextflow pipeline.nf --o "Arothron carduus" --t "genus" --docker "automagic" --outgroup "Cheilopogon agoo" --email "123abc@gmail.com" (the email address I put is just an example)
```

### Check Results

If outgroup was specified:
```
ls {organism}_{taxonomic level}_{outgroup}_Results/
```

If outgroup wasn't specified:
```
ls {organism}_{taxonomic level}_Results/
```

### Expected Outputs

- organism_seqs: Directory containing every gene sequences of every organism related to the one specified inside the specified taxonomic level + outgroup if specified;
- aln_data: Directory containing the aligned sequences previously saved;
- concatenated_seqs.fasta: FASTA file with every aligned sequences concatenated;
- model.out: Modeltest-NG output file conatining the results for the best evolutionary model;
- output.raxml.bestTree: Best phylogenetic tree file generated by RAxML-NG;
- output.raxml.bootstraps: Bootstrap trees file generated by RAxML-NG;
- output.raxml.bestTree.support: Support tree file generated by RAxML-NG;
- phylo_tree.svg: Phylogenetic tree file in SVG format.

## References
https://www.docker.com/

https://msa.sbc.su.se/cgi-bin/msa.cgi

https://www.nextflow.io/

https://github.com/amkozlov/raxml-ng

https://github.com/marekborowiec/AMAS

https://github.com/ddarriba/modeltest

https://toyplot.readthedocs.io/en/stable/

https://github.com/eaton-lab/toytree



